opencog/rule-engine/forwardchainer/ForwardChainer.cc

```diff
@@ -144,19 +149,51 @@ void ForwardChainer::do_chain(ForwardChainerCallBack& fcb,
      init_sources = _as.get_outgoing(hsource);
     else
         init_sources.push_back(hsource);
+
     _fcmem.update_potential_sources(init_sources);
 
-    _fcmem.set_source(fcb.choose_next_source(_fcmem)); //set initial source
     auto max_iter = _configReader.get_maximum_iterations();
 
     while (_iteration < max_iter /*OR other termination criteria*/) {
+
         _log->info("Iteration %d", _iteration);
 
-        do_step(fcb);
+        if(_rule_weight.empty())
+        {
+            //! Choose next source.
+            _log->info("[ForwardChainer] setting next source");
+
+            _fcmem.set_source(choose_next_source(_fcmem));
+
+            // Choose matching rules whose input matches with the source.
+            vector<Rule*> matched_rules = choose_rules(_fcmem);
+
+            //! If no rules matches the pattern of the source,
+            //! set all rules for candidacy to be selected by the proceeding step.
+            //! xxx this decision is made based on recent discussion.
+            //! it might still face some changes.
```

Comment:
What's the thinking here for why to do that (set all rules for candidacy if none match with the selected source)? Do we really want to do that?